[FEATURE BRANCH] Reduce the size of managedFields

jennybuckley opened the pull request:
**What type of PR is this?**
/kind feature

**Old Format:**
``` yaml
foo:
  apiVersion: v1
  fields:
    members:
    - fieldName: apiVersion
    - fieldName: kind
    children:
    - pathElement:
        fieldName: metadata
      set:
        members:
        - fieldName: name
        children:
        - pathElement:
            fieldName: labels
          set:
            members:
            - fieldName: app
    - pathElement:
        fieldName: spec
      set:
        members:
        - fieldName: replicas
        children:
        - pathElement:
            fieldName: selector
          set:
            children:
            - pathElement:
                fieldName: matchLabels
              set:
                members:
                - fieldName: app
        - pathElement:
            fieldName: template
          set:
            children:
            - pathElement:
                fieldName: metadata
              set:
                children:
                - pathElement:
                    fieldName: labels
                  set:
                    members:
                    - fieldName: app
            - pathElement:
                fieldName: spec
              set:
                children:
                - pathElement:
                    fieldName: containers
                  set:
                    members:
                    - key:
                      - name: name
                        value:
                          stringValue: nginx
                          'null': false
                    children:
                    - pathElement:
                        key:
                        - name: name
                          value:
                            stringValue: nginx
                            'null': false
                      set:
                        members:
                        - fieldName: image
                        - fieldName: name
                        children:
                        - pathElement:
                            fieldName: ports
                          set:
                            children:
                            - pathElement:
                                index: 0
                              set:
                                members:
                                - fieldName: containerPort
```

**New Format:**
``` yaml
foo:
  apiVersion: v1
  fields:
    f:apiVersion: {}
    f:kind: {}
    f:metadata:
      f:labels:
        f:app: {}
      f:name: {}
    f:spec:
      f:replicas: {}
      f:selector:
        f:matchLabels:
          f:app: {}
      f:template:
        f:medatada:
          f:labels:
            f:app: {}
        f:spec:
          f:containers:
            k:{"name":"nginx"}:
              .: {}
              f:image: {}
              f:name: {}
              f:ports:
                i:0:
                  f:containerPort: {}
```

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

EDIT: a lot of stuff got added onto this PR

Commit:
Encode/Decode PathElement to/from strings

Commit:
Add Field to Set conversion functions (apelisse)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply/roundtrip_test.go:
```diff
@@ -37,199 +37,93 @@ func TestRoundTripManagedFields(t *testing.T) {
+    i:5:
+      f:i: {}
+    v:3:
+      f:alspPi: {}
```
"alsp"

Is there a missing test? :) :)

jennybuckley commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/fuzzer/fuzzer.go:
```diff
@@ -276,14 +276,6 @@ func v1FuzzerFuncs(codecs runtimeserializer.CodecFactory) []interface{} {
-			// We have to customize the randomization of VersionedFieldSet because
-			// there is not currently a MaxDepth in the vendored version of gofuzz
-			// TODO (jennybuckley): remove this exception once we vendor in
-			// this PR https://github.com/google/gofuzz/pull/21
```
Can remove this, since https://github.com/kubernetes/kubernetes/commit/0c9660a45825b7d4d8131bbd139a046459cfdeb8 is in

Commit:
Refactor FieldManager code into single class

Commit:
Finish implementation

Commit:
Update generated

Commit:
Remove commented file that shouldn't be commeted

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply/decode.go:
```diff
@@ -43,131 +40,10 @@ func DecodeManagedFields(encodedManagedFields map[string]metav1.VersionedFieldSe
+	versionedSet.Set = &set
 	return versionedSet, nil
 }
-
```
That's a lot of removed code :+1:

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/apply/fields.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// FieldsToSet creates a set paths from an input trie of fields
+func FieldsToSet(f metav1.Fields) (fieldpath.Set, error) {
```
I suspect we should make these two methods private. But also there are other things that have a different level of "privateness" (`removeUselessDots` below should be private to only this file). I suspect we might want an internal package in here.

jennybuckley replied:
maybe  "apply/serializer"?

apelisse replied:
Most of what `apply` does it conversion from kubernetes <-> smd, so everything is serializer?

jennybuckley replied:
Or were you thinking of keeping encoder/decoder in the top level apply package

apelisse reviewed (commented):
We should probably re-organize this package a little bit, but that's out of scope for this PR.
I'm very happy with the format, I think we can do what Daniel suggests and remove the "f:" under some conditions, but it should be backward compatible anyway, so I think we can tackle that later (not that it would take a long time either).
/lgtm

apelisse commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -1046,77 +1046,12 @@ type VersionedFieldSet struct {
 	APIVersion string `json:"apiVersion,omitempty" protobuf:"bytes,1,opt,name=apiVersion"`
 	// Fields identifies a set of fields.
-	Fields FieldSet `json:"fields,omitempty" protobuf:"bytes,2,opt,name=fields,casttype=FieldSet"`
+	Fields Fields `json:"fields,omitempty" protobuf:"bytes,2,opt,name=fields,casttype=Fields"`
```
Do we want to make it `omitempty` or do we want to keep the ability to differentiate between completely removed and set-to-null/empty? We could also differentiate based on the lack of "VersionedFieldSet" vs lack of Fields.

Commit:
Add managed fields on create

Commit:
Refactor

Commit:
Fix rebase

Commit:
Add owners for feature branch only

jennybuckley commented:
Rebased this on  #72066

apelisse commented:
/priority important-longterm

Commit:
Remove managed fields before we set them

That's going to prevent recursively inserting managed fields in managed
fields...

Commit:
Make server side apply flag consistent with dry-run

Commit:
Add --force-conflicts flag to kubectl

The flag allows force on apply so that you can re-acquire ownership of
fields when you have a conflict.

Commit:
Fix api and update vendored smd

apelisse commented:
/lgtm

Commit:
add reminder to remove approvers

apelisse commented:
/lgtm

jennybuckley commented:
/retest
